reactos/drivers/usb/usbohci_new/usbohci.c

```diff
@@ -0,0 +1,2752 @@
+    DPRINT_OHCI("OHCI_OpenInterruptEndpoint: InitTD. Index[PeriodIdx] - %x, ScheduleOffset - %x\n",
+                Index[PeriodIdx],
+                ScheduleOffset);
+
+    OhciEndpoint->HeadED = &OhciExtension->IntStaticED[Index[PeriodIdx] +
+                                                       ScheduleOffset];
+
+    //OhciEndpoint->HeadED->UsbBandwidth += EndpointProperties->UsbBandwidth;
+
+    OHCI_InitializeTDs(OhciEndpoint, EndpointProperties);
+
+    OhciEndpoint->HcdED = OHCI_InitializeED(OhciEndpoint,
+                                            ED,
+                                            OhciEndpoint->FirstTD,
+                                            EndpointProperties->BufferPA);
+
+    OHCI_InsertEndpointInSchedule(OhciEndpoint);
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+OHCI_OpenIsoEndpoint(IN POHCI_EXTENSION OhciExtension,
+                     IN PUSBPORT_ENDPOINT_PROPERTIES EndpointProperties,
+                     IN POHCI_ENDPOINT OhciEndpoint)
+{
+    DPRINT1("OHCI_OpenIsoEndpoint: UNIMPLEMENTED. FIXME\n");
+    return MP_STATUS_NOT_SUPPORTED;
+}
+
+MPSTATUS
+NTAPI
+OHCI_OpenEndpoint(IN PVOID ohciExtension,
+                  IN PVOID endpointParameters,
+                  IN PVOID ohciEndpoint)
+{
+    POHCI_EXTENSION OhciExtension = ohciExtension;
+    POHCI_ENDPOINT OhciEndpoint = ohciEndpoint;
+    PUSBPORT_ENDPOINT_PROPERTIES EndpointProperties = endpointParameters;
+    ULONG TransferType;
+    MPSTATUS MPStatus;
+
+    DPRINT_OHCI("OHCI_OpenEndpoint: ... \n");
+
+    RtlCopyMemory(&OhciEndpoint->EndpointProperties,
+                  endpointParameters,
+                  sizeof(USBPORT_ENDPOINT_PROPERTIES));
+
+    InitializeListHead(&OhciEndpoint->TDList);
+
+    TransferType = EndpointProperties->TransferType;
+
+    switch (TransferType)
+    {
+        case USBPORT_TRANSFER_TYPE_ISOCHRONOUS:
+            MPStatus = OHCI_OpenIsoEndpoint(OhciExtension,
+                                            EndpointProperties,
+                                            OhciEndpoint);
+            break;
+
+        case USBPORT_TRANSFER_TYPE_CONTROL:
+            MPStatus = OHCI_OpenControlEndpoint(OhciExtension,
+                                                EndpointProperties,
+                                                OhciEndpoint);
+            break;
+
+        case USBPORT_TRANSFER_TYPE_BULK:
+            MPStatus = OHCI_OpenBulkEndpoint(OhciExtension,
+                                             EndpointProperties,
+                                             OhciEndpoint);
+            break;
+
+        case USBPORT_TRANSFER_TYPE_INTERRUPT:
+            MPStatus = OHCI_OpenInterruptEndpoint(OhciExtension,
+                                                  EndpointProperties,
+                                                  OhciEndpoint);
+            break;
+
+        default:
+            MPStatus = MP_STATUS_NOT_SUPPORTED;
```

Comment:
DPRINT1 TransferType ?

Reply:
If we are here, then it will be critical, since there are no other types of transfers.